### PR TITLE
Add Accordion Web Component

### PR DESCRIPTION
**Pull request by bkjohnson**
## Description
Closes https://github.com/department-of-veterans-affairs/va.gov-team/issues/18079

This adds two Web Components:

- `<va-accordion>`
    - A wrapper component for the item. Controls some item styling as well as closing other items when one is opened
- `<va-accordion-item>`
    - A collapsible item containing hidden content.

## Testing done

- Local storybook testing
- Testing in `vets-website`.
- e2e tests written

### From vets-website

component gif

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

File: package.json
Comment (bkjohnson) [inline package.json:33]: Right now I'm using `postcss` to inline the SVG background images. Coincidentally, it also inlines the `@import` rules.

File: src/components/va-accordion/va-accordion-item.css
Comment (bkjohnson) [inline src/components/va-accordion/va-accordion-item.css:1]: `postcss` will inline this. Using imports like this can allow us to have shared styling across components, since this normally doesn't work due to the shadow DOM.

File: src/components/va-accordion/va-accordion-item.css
Comment (bkjohnson) [inline src/components/va-accordion/va-accordion-item.css:8]: The CSS for `<va-accordion>` sets these based on the `bordered` prop/attribute.

File: src/components/va-accordion/va-accordion.stories.tsx
Comment (bkjohnson) [inline src/components/va-accordion/va-accordion.stories.tsx:25]: These stories right now are only used for local development. We may explore exporting these and importing them into the storybook in `master` so that we don't have to have duplicate code.

File: src/global/variables.css
Comment (bkjohnson) [inline src/global/variables.css:7]: This file contains CSS variables that can be used across Components. This is how we can style certain shadow DOM things at a global level.

File: src/mixins/focusable.css
Comment (bkjohnson) [inline src/mixins/focusable.css:4]: From [USWDS styles](https://github.com/uswds/uswds/blob/3dc296ec56cd621fe52d918701fd94621d96a198/src/stylesheets/core/mixins/_focus.scss#L12-L13).

File: src/components/va-accordion/va-accordion.tsx
Comment (cvalarida) [inline src/components/va-accordion/va-accordion.tsx:17]: Is it standard practice to yell the slot name? :sweat_smile:

File: src/utils/utils.ts
Comment (cvalarida) [inline src/utils/utils.ts:7]: 1) Can this be `querySelector` to return the first node?
2) What happens if there are none?

**Review by cvalarida — COMMENTED** [review]
This looks pretty good despite my predilection of leaving a million comments. :sweat_smile:

The two main things I'm holding off on approving are:
1. Using `aria-expanded` for logic instead of "display" (to screen readers)
2. Using the `:host` CSS selector in IE (I just don't know if it'll work)

Other than that, I mostly just had a bunch of questions to help me understand web components better.

File: src/assets/minus.svg
Comment (bkjohnson) [reply src/assets/minus.svg:2]: Oops! I meant to link this in the PR, but I _did_ copy this from somewhere - the [USWDS file](https://github.com/uswds/uswds/blob/39f62c0eb7fc0a2de85d1d942cbea43dc1832f99/src/img/minus.svg).

File: src/assets/plus.svg
Comment (bkjohnson) [reply src/assets/plus.svg:2]: Copied from the [USWDS file](https://github.com/uswds/uswds/blob/39f62c0eb7fc0a2de85d1d942cbea43dc1832f99/src/img/plus.svg).

File: src/components.d.ts
Comment (bkjohnson) [reply src/components.d.ts:35]: This file is auto-generated, I'll make that change in the `.tsx` file.

File: src/components.d.ts
Comment (bkjohnson) [reply src/components.d.ts:25]: Yeah - I'll make that comment a bit clearer.

File: src/components/va-accordion/va-accordion-item.tsx
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion-item.tsx:35]: Correct.

File: src/components/va-accordion/va-accordion-item.css
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion-item.css:22]: Yeah, PostCSS inlines the url:

image

File: src/components.d.ts
Comment (bkjohnson) [reply src/components.d.ts:23]: This is auto-generated :shrug:

File: src/components/va-accordion/va-accordion-item.css
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion-item.css:3]: From [Stencil's docs on Styling & IE Support](https://stenciljs.com/docs/styling#ie-support):

> - Global CSS variables can only be declared in `:root` or `html`, they can't be dynamic.
> - CSS variables within a component can ONLY be defined within a `:host(...)` selector.
> - CSS variables within a component can be consumed (`var(--thing)`) in any selector.

So, this will work in IE.

File: src/components/va-accordion/va-accordion.tsx
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion.tsx:17]: :laughing: This is why I did that:

<details>
    <summary>The diff to get the log</summary>
    
    ```diff
    diff --git a/src/utils/utils.ts b/src/utils/utils.ts
index 9a794c1..8826330 100644
--- a/src/utils/utils.ts
+++ b/src/utils/utils.ts
@@ -3,6 +3,7 @@ export function format(first: string, middle: string, last: string): string {
 }
 
 export function getSlottedNodes(root: HTMLElement, nodeName: string): Array<Node> {
+  console.log(root.shadowRoot.querySelectorAll('slot')[0].assignedNodes());
   return root.shadowRoot
     .querySelectorAll('slot')[0]
     .assignedNodes()
     ```
     
 </details>
 
image


 I'll tweak the util function though so that it compares things that have been lowercased.

File: src/utils/utils.ts
Comment (bkjohnson) [reply src/utils/utils.ts:7]: If there are no `<slot>`s, then this function is being used somewhere that it shouldn't be. I'll leave a comment on it, but it is explicitly meant to be used to return the nodes inside of a `<slot>` matching `nodeName`, all inside a given root node.

File: src/components/va-accordion/va-accordion-item.e2e.ts
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion-item.e2e.ts:4]: It's a bit blurry to me too :thinking:

File: src/components/va-accordion/va-accordion-item.css
Comment (cvalarida) [reply src/components/va-accordion/va-accordion-item.css:39]: Hrm. The more I try to figure out how to explain this, the less I'm convinced I'm right. Seemed weird to tie the visual styles with the `aria-` attributes, but maybe that makes the most sense after all. :man_shrugging:

**Review by cvalarida — COMMENTED** [review]
Eh, I guess never mind on the whole `aria-` attribute thing from my last review. I don't really have a strong argument. If it comes back to bite us in a strange way (like screen readers weirdly manipulating the `aria-expanded` attribute), we'll...well, we'll probably never know, if we're honest, but that's probably...fine?

Let's figure out how to differentiate unit tests with e2e tests before we call this a-go, though.

File: src/components/va-accordion/va-accordion-item.e2e.ts
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion-item.e2e.ts:4]: I think the testing is better focused now.

File: package.json
Comment (bkjohnson) [inline package.json:18]: Building this json file will make it easy to include component documentation in our main Storybook instance.

File: src/components/va-accordion/va-accordion-item.e2e.ts
Comment (cvalarida) [inline src/components/va-accordion/va-accordion-item.e2e.ts:21]: Should we also test that the item contents are expanded, or is that...just kinda always on (hidden only by CSS)?

I don't suppose you know of a way to test the display of the content, eh? Like, the `display` CSS property. Because while this tests the `aria-expanded` attribute, it makes the assumption that the styling for that attribute _does_ something, y'know? That's probably way out of scope for this PR, though. :man_shrugging:

File: src/components/va-accordion/va-accordion-item.e2e.ts
Comment (cvalarida) [inline src/components/va-accordion/va-accordion-item.e2e.ts:35]: Okay, so combine this with the above test, and we're ensuring that clicking fires the event, and that when the `open` property is set to `true`, it'll be opened like we expect. The part we're missing is where we stitch the two together, but that happens in the `va-accordion` wrapper.

I'm following. :+1: Might be good to make a comment either here or in the component itself saying what the purpose is of firing the event.

File: src/components/va-accordion/va-accordion.e2e.ts
Comment (cvalarida) [inline src/components/va-accordion/va-accordion.e2e.ts:32]: Maybe also test that `buttons[1]`'s `aria-expanded` is still false.

File: src/components/va-accordion/va-accordion.e2e.ts
Comment (cvalarida) [inline src/components/va-accordion/va-accordion.e2e.ts:55]: Same here.

File: src/components/va-accordion/va-accordion.tsx
Comment (cvalarida) [inline src/components/va-accordion/va-accordion.tsx:66]: I love this simplicity. It's beautiful.

File: src/global/variables.css
Comment (cvalarida) [inline src/global/variables.css:1]: I don't know what the answer to this is yet, but we should document the difference between these variables, what's in `formation`, and what's on `vets-website`. Not that there necessarily is a difference in values, but that these values live in potentially three different places.

Bonus points for dropping a comment in here with a link to where you got these values.

File: src/mixins/focusable.css
Comment (cvalarida) [reply src/mixins/focusable.css:4]: Bonus points for leaving this as a comment for easy traceability.

**Review by cvalarida — APPROVED** [review]
This looks great! I'm ready and willing to give extra credit for dropping some more comments in the code for future maintainers. :grin:

File: src/components/va-accordion/va-accordion-item.e2e.ts
Comment (bkjohnson) [reply src/components/va-accordion/va-accordion-item.e2e.ts:21]: I would rather test that using screenshot diffs than test for the presence of CSS properties. That seems....awkward.

**Pull request merged**
